File: scripts/extract_mbtiles.js
#!/usr/bin/env node

/**
 * Extract tiles from MBTiles file to folder structure
 * Uses better-sqlite3 for efficient streaming
 */

const fs = require('fs');
const path = require('path');
const Database = require('better-sqlite3');

function extractMBTiles(mbtilesPath, outputDir) {
  if (!fs.existsSync(mbtilesPath)) {
    console.error(`Error: File not found: ${mbtilesPath}`);
    return false;
  }

  console.log(`Opening MBTiles file: ${mbtilesPath}`);
  console.log('This may take a while for large files...\n');

  try {
    // Open database in readonly mode
    const db = new Database(mbtilesPath, { readonly: true, fileMustExist: true });

    // Get tile count
    const countStmt = db.prepare("SELECT COUNT(*) as count FROM tiles");
    const totalTiles = countStmt.get().count;
    console.log(`Found ${totalTiles.toLocaleString()} tiles to extract`);
    console.log(`Output directory: ${outputDir}\n`);

    // Prepare statement for iterating tiles
    const stmt = db.prepare("SELECT zoom_level, tile_column, tile_row, tile_data FROM tiles");
    
    let extracted = 0;
    let lastPercent = -1;

    // Stream tiles one at a time
    for (const row of stmt.iterate()) {
      const zoom = row.zoom_level;
      const x = row.tile_column;
      const yTms = row.tile_row;
      const tileData = row.tile_data;

      // Convert TMS y to XYZ y (flip vertically)
      const y = Math.pow(2, zoom) - 1 - yTms;

      // Create directory structure
      const tileDir = path.join(outputDir, String(zoom), String(x));
      if (!fs.existsSync(tileDir)) {
        fs.mkdirSync(tileDir, { recursive: true });
      }

      // Write tile
      const tilePath = path.join(tileDir, `${y}.png`);
      fs.writeFileSync(tilePath, tileData);

      extracted++;
      const percent = Math.floor((extracted / totalTiles) * 100);
      if (percent !== lastPercent && (extracted % 1000 === 0 || percent % 5 === 0)) {
        process.stdout.write(`\rProgress: ${extracted.toLocaleString()}/${totalTiles.toLocaleString()} (${percent}%)`);
        lastPercent = percent;
      }
    }

    console.log(`\n\n✅ Successfully extracted ${extracted.toLocaleString()} tiles!`);
    db.close();
    return true;

  } catch (error) {
    console.error('Error during extraction:', error.message);
    if (error.message.includes('Could not find any Visual Studio')) {
      console.error('\nThis module requires compilation but Visual Studio Build Tools are not installed.');
      console.error('Please use the pre-extracted tiles approach instead.');
    }
    return false;
  }
}

async function main() {
  console.log('='.repeat(60));
  console.log('MBTiles Extractor');
  console.log('='.repeat(60));
  console.log();

  if (process.argv.length < 3) {
    console.log('Usage: node extract_mbtiles.js <path-to-mbtiles-file>');
    console.log('');
    console.log('Example:');
    console.log('  node extract_mbtiles.js C:\\Downloads\\map.mbtiles');
    console.log('');
    process.exit(1);
  }

  const mbtilesPath = process.argv[2];

  // Output to tiles directory
  const outputDir = path.join(__dirname, '..', 'tiles');

  const success = extractMBTiles(mbtilesPath, outputDir);

  console.log();
  console.log('='.repeat(60));
  if (success) {
    console.log("Done! Run 'npm run check-tiles' to verify.");
  } else {
    console.log('Extraction failed. Please check the error messages above.');
  }
  console.log('='.repeat(60));

  process.exit(success ? 0 : 1);
}

main();

File: scripts/extract_mbtiles.py
#!/usr/bin/env python
"""
Extract tiles from MBTiles file to folder structure
No external dependencies required - uses built-in sqlite3
"""

import sqlite3
import os
import sys

def extract_mbtiles(mbtiles_path, output_dir):
    """Extract tiles from MBTiles to {z}/{x}/{y}.png structure"""
    
    if not os.path.exists(mbtiles_path):
        print(f"Error: File not found: {mbtiles_path}")
        return False
    
    print(f"Opening MBTiles file: {mbtiles_path}")
    
    try:
        # Connect to MBTiles database
        conn = sqlite3.connect(mbtiles_path)
        cursor = conn.cursor()
        
        # Get tile count
        cursor.execute("SELECT COUNT(*) FROM tiles")
        total_tiles = cursor.fetchone()[0]
        print(f"Found {total_tiles:,} tiles to extract")
        print(f"Output directory: {output_dir}")
        print()
        
        # Extract tiles
        cursor.execute("SELECT zoom_level, tile_column, tile_row, tile_data FROM tiles")
        
        extracted = 0
        for row in cursor:
            zoom, x, y_tms, tile_data = row
            
            # Convert TMS y to XYZ y (flip vertically)
            y = (2 ** zoom - 1) - y_tms
            
            # Create directory structure
            tile_dir = os.path.join(output_dir, str(zoom), str(x))
            os.makedirs(tile_dir, exist_ok=True)
            
            # Write tile
            tile_path = os.path.join(tile_dir, f"{y}.png")
            with open(tile_path, 'wb') as f:
                f.write(tile_data)
            
            extracted += 1
            if extracted % 100 == 0:
                percent = (extracted / total_tiles) * 100
                print(f"\rProgress: {extracted:,}/{total_tiles:,} ({percent:.1f}%)", end='', flush=True)
        
        print(f"\n✅ Successfully extracted {extracted:,} tiles!")
        conn.close()
        return True
        
    except sqlite3.Error as e:
        print(f"Database error: {e}")
        return False
    except Exception as e:
        print(f"Error: {e}")
        return False

if __name__ == "__main__":
    print("=" * 60)
    print("MBTiles Extractor")
    print("=" * 60)
    print()
    
    if len(sys.argv) < 2:
        print("Usage: python extract_mbtiles.py <path-to-mbtiles-file>")
        print()
        print("Example:")
        print("  python extract_mbtiles.py C:\\Downloads\\map.mbtiles")
        print()
        sys.exit(1)
    
    mbtiles_path = sys.argv[1]
    
    # Output to tiles directory
    script_dir = os.path.dirname(os.path.abspath(__file__))
    output_dir = os.path.join(os.path.dirname(script_dir), 'tiles')
    
    success = extract_mbtiles(mbtiles_path, output_dir)
    
    print()
    print("=" * 60)
    if success:
        print("Done! Run 'npm run check-tiles' to verify.")
    else:
        print("Extraction failed. Please check the error messages above.")
    print("=" * 60)
    
    sys.exit(0 if success else 1)
